src/functional/sim_piece.zig

```diff
@@ -0,0 +1,765 @@
+fn appendSegmentMetadata(
+    metadata_map: *HashMap(
+        f64,
+        ArrayList(SegmentMetadata),
+        HashF64Context,
+        std.hash_map.default_max_load_percentage,
+    ),
+    metadata: SegmentMetadata,
+    allocator: mem.Allocator,
+) !void {
+    const get_result = try metadata_map.getOrPut(metadata.interception);
+    if (!get_result.found_existing) {
+        get_result.value_ptr.* = ArrayList(SegmentMetadata).init(allocator);
+    }
+    try get_result.value_ptr.*.append(metadata);
+}
+
+/// Returns a comparator function that compares SegmentMetadata by `lower_bound_slope`.
+fn compareMetadataBySlope() fn (void, SegmentMetadata, SegmentMetadata) bool {
+    return struct {
+        pub fn inner(_: void, metadata_one: SegmentMetadata, metadata_two: SegmentMetadata) bool {
+            return metadata_one.lower_bound_slope < metadata_two.lower_bound_slope;
+        }
+    }.inner;
+}
+
+/// Returns a comparator function that compares SegmentMetadata by `start_time`.
+fn compareMetadataByStartTime() fn (void, SegmentMetadata, SegmentMetadata) bool {
+    return struct {
+        pub fn inner(_: void, metadata_one: SegmentMetadata, metadata_two: SegmentMetadata) bool {
+            return metadata_one.start_time < metadata_two.start_time;
+        }
+    }.inner;
+}
+
+/// Append `value` of `type` determined at compile time to `compressed_values`.
+fn appendValue(comptime T: type, value: T, compressed_values: *std.ArrayList(u8)) !void {
+    // Compile-time type check
+    switch (@TypeOf(value)) {
+        f64, usize => {
+            const value_as_bytes: [8]u8 = @bitCast(value);
+            try compressed_values.appendSlice(value_as_bytes[0..]);
+        },
+        f32 => {
+            const value_as_bytes: [4]u8 = @bitCast(value);
+            try compressed_values.appendSlice(value_as_bytes[0..]);
+        },
+        else => @compileError("Unsupported type for append value function"),
+    }
+}
+
+test "f64 context can hash" {
+    const allocator = testing.allocator;
+    var f64_hash_map = HashMap(
+        f64,
+        f64,
+        HashF64Context,
+        std.hash_map.default_max_load_percentage,
+    ).init(allocator);
+    defer f64_hash_map.deinit();
+    var rnd = std.rand.DefaultPrng.init(@as(u64, @bitCast(std.time.milliTimestamp())));
+
+    // Even a small deviation like this should not impact the ability to hash the numbers.
+    const deviation_margin = 1e-16;
+    for (0..100) |_| {
+        const rand_number = rnd.random().float(f64) - 0.5;
+        try f64_hash_map.put(rand_number, rand_number);
+        try f64_hash_map.put(rand_number - deviation_margin, rand_number - deviation_margin);
+        try f64_hash_map.put(rand_number + deviation_margin, rand_number + deviation_margin);
+    }
+
+    // All elements are expected to be hashed independently, without collisions.
+    try testing.expectEqual(300, f64_hash_map.count());
```

Comment:
How does `count()` if the values does not collide? If this is done maybe extend the comment so it explains how, if it is not done maybe delete the comment or revise it.

Reply:
I think the comment now explains the logic.

Reply:
It may just be me that is missing something, but it is still not clear to me how the use of `count()` can be used to check for [hash collisions](https://en.wikipedia.org/wiki/Hash_table#Collision_resolution). Do you mean that the values should all be stored as separate key-value pairs in the hash map and not that hash collisions have occurred?

Reply:
Yes, that is exactly what I mean. Changed the comment to make it even clearer.